### ISSUE TITLE
reuse codehash->address caches across multiple DSProxies

### ISSUE THREAD

**Issue by nmushegian** [dapphub/ds-proxy#2]
e.g. in their factory

**Comment by nmushegian** [comment]
Looks like there is no caching at all yet. That would be a good place to start or else we will really pollute the chain

**Comment by NiklasKunkel** [comment]
Can you elaborate on what you mean by this? This sounds similar to ds-robot except for the multiple DSProxies part.

**Comment by nmushegian** [comment]
That explains my confusion. This should be an issue for DSRobot: suggestion to use factory to cache libs

**Comment by nmushegian** [comment]
wait no:  even in DSProxy, you should have a codehash -> deployment cache, to avoid redundant creates

**Comment by NiklasKunkel** [comment]
Update, I've created a new branch for the global cache (something I should have done a while ago...). The current commit on master doesn't work so I may roll back master to the last working commit. 

After consulting with @nmushegian I've moved the cache to be inside the Proxy Factory rather than the Proxy instance. This makes more sense, all proxys can utilizie the same cache. However for flexibility I've left in the option for a proxy instance to change its cache. This is important because storage costs do not scale linearly. Given a sufficiently large cache, a user may want to avoid having to pay storage costs to cache a new contract in the global cache. 

I'm having an issue with DS-Auth at the moment where a setOwner(msg.sender) is failing that is prohibiting me from continuing to test. It's strange that it seems to work fine under "dapp test" conditions but fails on Kovan. Will create a new issue for this.
